mdbook/src/15-interrupts/sharing-data-with-globals.md

```diff
@@ -0,0 +1,282 @@
+handler runs and also tries to update `COUNTER`. The usual chaos will ensue. Clearly some kind of
+locking is in order.
+
+The `critical-section` crate provides a sort of `Mutex` type, but with an unusual API and unusual
+operations. Examining the `Cargo.toml` for this chapter, you will see the feature
+`critical-section-single-core` on the `cortex-m` crate enabled. This feature asserts that there is
+only one processor core in this system, and that thus synchronization can be performed by simply
+*disabling interrupts* across the critical section. If not in an interrupt, this will ensure that
+only the main program has access to the global. If in an interrupt, this will ensure that the main
+program cannot be accessing the global (program control is in the interrupt handler) and that no
+other higher-priority interrupt handler can fire.
+
+`critical_section::Mutex` is a bit weird in that it gives mutual exclusion but does not itself give
+mutability. To make the data mutable, you will need to protect an interior-mutable type — usually
+`RefCell` — with the mutex. This `Mutex` is also a bit weird in that you don't `.lock()`
+it. Instead, you initiate a critical section with a closure that receives a "critical section token"
+certifying that other program execution is prevented. This token can be passed to the `Mutex`'s
+`borrow()` method to allow access.
+
+Putting it all together gives you the ability to share state between ISRs and the main program
+(`examples/count-once.rs`).
+
+```rust
+{{#include examples/count-once.rs}}
+```
+
+You still cannot safely return from your ISR, but now you are in a position to do something about
+that: share the `GPIOTE` with the ISR so that the ISR can clear the interrupt.
+
+### Sharing Peripherals (etc) With Globals
+
+There's one more problem yet to solve: Rust globals must be initialized statically — before the
+program starts. For the counter that was easy — just initialize it to 0. If you want to share the
+`GPIOTE` peripheral, though, that won't work. The peripheral must be retrieved from the `Board`
+struct and set up once the program has started: there is no `const` initializer for this (nor can
+there reasonably be).
+
+Let's rewrite the button counter a bit. First, move the actual count to be an `AtomicUsize`. This is
+a more natural type for this global anyhow. Next, add a global `GPIOTE_PERIPHERAL` variable using
+the `LockMut` type from the `critical-section-lock-mut` crate. This crate is a convenient wrapper
+for the pattern of the last section.
+
+Now that the main program can set up the GPIOTE peripheral and then make it available to the
+interrupt handler, you can quit panicking and let the counter bump up on every button press. Move
+the count display into the main loop, to show that the count is shared between the interrupt handler
+and the rest of the program.
+
+Give this example (`examples/count.rs`) a run and note that the count is bumped up 1 on every push
```

Comment:
Since `examples/count.rs` is not included in the text, it would be nice to provide a link to it.